### ISSUE TITLE
FoLiA-abby: linebreaks

### ISSUE THREAD

**Issue by pirolen** [LanguageMachines/foliautils#53]
If possible, when there is a new line in the Abbyy XML, it would be great to represent them with a corresponding linebreak element in FoLiA.

**Comment by kosloot** [comment]
I implemented an --addbreaks option in FoLiA-abby.

**BUT:** I also refactored FoLiA-abby to create a more expressive file, with nested ``<part>`` nodes.
Outer ``<part>`` nodes have ``class="line"``. Inner nodes ``class="fragment".``
One line can have several "fragment"'s
A newline (``<br>``) is in fact implicit after every "line".

I assume (hope) this is more clear and versatile. @pirolen maybe you can check this out?
Hopefully this also solves most of the spurious spacing in the previous implementation.

**Comment by pirolen** [comment]
Thanks! Looking fwd to testing this, but my VM is down right now :-(
Meanwhile I'm observing items in the previous version.
Could you perhaps check what happens in the attached files (abbyy and FoLiA xmls) in this (abbyy xml) line:

```
<par leftIndent="2100" startIndent="-2100" lineSpacing="1171" style="{2FB57E6D-73D6-4054-9499-28923DAC219E}">
<line baseline="359" l="152" t="337" r="1086" b="366"><formatting lang="FrenchStandard" ff="Times New Roman" fs="10." style="{293DE961-0710-4439-9CBF-3D26B0D6AD0D}">
<charParams l="152" t="342" r="166" b="358">A</charParams>
```

In the FoLiA, the linebreak is put after the A: 
`A <br/>10 `
but the linebreak is before it -- it is between two "parts" with different formatting. So perhaps the break is supposed to have its own part? Perhaps you could check how the new converter handles it?

[b1_23_p24.folia.xml.txt](https://github.com/LanguageMachines/foliautils/files/5913215/b1_23_p24.folia.xml.txt)
[b1_23_p24.xml.txt](https://github.com/LanguageMachines/foliautils/files/5913216/b1_23_p24.xml.txt)

**Comment by proycon** [comment]
Sorry I kind of missed the discussion that introduced the 'parts', but I have an important comment on it here:  https://github.com/LanguageMachines/ucto/issues/80#issuecomment-771832599

TL;DR: I don't think we should use parts here

**Comment by kosloot** [comment]
yes, it seems I took the the wrong path there. So I have to rethink the rethinking.

**Comment by kosloot** [comment]
So: at the moment we have the incorrect: (constructed example)
```xml
      <p xml:id="FA-piroska.text.div1.p5">
        <feat class="{2C092230-2651-425A-ACE2-047B8277C192}" subset="par_style"/>
        <part xml:id="FA-piroska.text.div1.p5.part.1" class="line">
          <part xml:id="FA-piroska.text.div1.p5.part.1.part.1" class="fragment">
            <t class="OCR">
              <t-style><feat class="Times New Roman" subset="font_family"/><feat class="15." subset="font_size"/><feat class="{96A59073-7007-4008-A803-6C3663B19E8A}" subset="font_style"/>IV. Die römische Landwirtschaft und die </t-style>
            </t>
          </part>
          <part xml:id="FA-piroska.text.div1.p5.part.1.part.2" class="fragment">
            <t class="OCR">
              <t-style><feat class="Arial" subset="font_family"/><feat class="7.5" subset="font_size"/><feat class="{EF179B17-CBAB-451F-95F9-D59D59B05037}" subset="font_style"/>A220<br/></t-style>
            </t>
          </part>
          <part xml:id="FA-piroska.text.div1.p5.part.2" class="line">
            <t class="OCR">
              <t-style><feat class="Times New Roman" subset="font_family"/><feat class="11." subset="font_size"/><feat class="{29589F1E-6010-4DF8-A0E8-D990639E17BE}" subset="font_style"/>der Rückblick, den wir auf die älteren Verhältnisse der römischen</t-style>
            </t>
          </part>
        </part>
      </p>
```
That then should become something like:

```xml
      <p xml:id="FA-piroska.text.div1.p5">
        <feat class="{2C092230-2651-425A-ACE2-047B8277C192}" subset="par_style"/>
        <t class="OCR">
          <t-str xml:id="FA-piroska.text.div1.p5.part.1" class="fragment">
             <t-style><feat class="Times New Roman" subset="font_family"/><feat class="15." subset="font_size"/><feat class="{96A59073-7007-4008-A803-6C3663B19E8A}" subset="font_style"/>IV. Die römische Landwirtschaft und die </t-style>
           </t-str>
           <t-str xml:id="FA-piroska.text.div1.p5.part.2" class="fragment">
             <t-style><feat class="Arial" subset="font_family"/><feat class="7.5" subset="font_size"/><feat class="{EF179B17-CBAB-451F-95F9-D59D59B05037}" subset="font_style"/>A220</t-style>
            </t-str>
            <br/>
            <t-str xml:id="FA-piroska.text.div1.p5.part.3" class="line">
              <t-style><feat class="Times New Roman" subset="font_family"/><feat class="11." subset="font_size"/><feat class="{29589F1E-6010-4DF8-A0E8-D990639E17BE}" subset="font_style"/>der Rückblick, den wir auf die älteren Verhältnisse der römischen</t-style>
           </t-str>
         </t>
      </p>
```
This seems doable. But quite a change....

And also, this example is incorrect, as it introduces a wealth of spurious spaces. as spaces are significant inside ``<t-style>`` 
So the FoLiA will become human-unreadable. but ok.

**Comment by pirolen** [comment]
> So the FoLiA will become human-unreadable.

Oh no... one huge plus of the Abby->FoLiA conversion is the human-readability.

**Comment by kosloot** [comment]
I will do my best to get both valid AND useful FoLiA. its is a pita though.

**Comment by pirolen** [comment]
Thanks very much! I keep looking at the output, in my dev lamachine, testing the linebreaks/spaces/compatibility with folia2text/folia2html/ucto...

**Comment by pirolen** [comment]
> In the FoLiA, the linebreak is put after the A:
> `A <br/>10 `
> but the linebreak is before it 

The new version of the converter does not make this error, great, thanks!

**Comment by kosloot** [comment]
I think that the new solution is fundamentally better, but there are a lot of small issues.
We have to make some choices, which might not suite everyone.

I should work out some examples, but  atm I am distracted a bit. I hope you can work with what's available now.
The issues are mainly  about the formatting. Like 'how to interpret ``<t-hbr>``' and 'which spaces are relevant'

**Comment by pirolen** [comment]
Sure. Do I get it right that the way the converter constructs the elements/classes is now finalized, but e.g. ucto and foliatools might not yet interpret them optimally?
If the elements/classes are stable then I could start referencing them in FLAT and foliapy (othewise I'd wait).

**Comment by pirolen** [comment]
Revisiting the soft breaks: a while ago it was agreed that if in Abbyy XML there is end-of-line hyphen, this gets converted to t-hbr. Just wondering if it is generic enough. Suppose someone would actually like to keep those hyphen characters, since there are at least two kinds: '-' and '¬'. Maybe the --addbreaks option could be configurable in this respect?

**Comment by kosloot** [comment]
It is easy to keep the value of the hyphen in the resulting FoLiA, like this:
```xml
<t-style><feat class="GermanStandard" subset="ocr_lang"/>Da in diesem Kapitel hauptsächlich eine Erschei<t-hbr>-</t-hbr></t-style>
<t-style><feat class="GermanStandard" subset="ocr_lang"/>Betnebsweise.</t-style>
```

and:
```xml
t-str xml:id="FA-piroska.text.div1.p7.t-str.9">
  <t-style><feat class="GermanStandard" subset="ocr_lang"/>römische Landwirtschaft, wie sie uns die scriptores rei rusticae schil<t-hbr>¬</t-hbr></t-style>
  <t-style><feat class="GermanStandard" subset="ocr_lang"/>dern.</t-style>
</t-str>
```
BUT: this would mean that FoLiA's text extracting functions would output those characters too:
```
Da in diesem Kapitel hauptsächlich eine Erschei- Betnebsweise.
```
and:
```
römische Landwirtschaft, wie sie uns die scriptores rei rusticae schil¬ dern,
```
Is that desirable?

**Comment by pirolen** [comment]
In some scenarios it is, but not always. I wonder if it could be made optional?

**Comment by kosloot** [comment]
In general NO, that would introduce all kind of text-consistency problems in FoLiA

What IS possible, is to make it optional in FoLiA-abby, like the add-breaks option. A --keephyphens option is easy to add.
It would  allow ``<t-hbr/>`` vs. ``<t-hbr>-</t-hbr>``

**Comment by pirolen** [comment]
Thanks, that is what I meant to say, to make it optional in FoLiA-abby; sorry if it was confusing.

**Comment by kosloot** [comment]
I added a --keephyphens option now

**Comment by pirolen** [comment]
Thank you, will start using it!
